File: typescript/src/integrations/llamaindex.ts
/**
 * LlamaIndex integration for AgentSentinel.
 *
 * Wraps LlamaIndex tools and query engines with AgentSentinel protection.
 *
 * @example
 * ```ts
 * import { AgentPolicy, AgentGuard } from "@agentsentinel/sdk";
 * import { LlamaIndexGuard, protectLlamaIndexAgent } from "@agentsentinel/sdk/integrations/llamaindex";
 *
 * const policy = new AgentPolicy({ dailyBudget: 10.0 });
 * const guard = new AgentGuard({ policy });
 * const llamaGuard = new LlamaIndexGuard(guard);
 *
 * const queryKb = llamaGuard.tool(
 *   (query: string) => `Results for ${query}`,
 *   { name: "query_knowledge_base", model: "gpt-4o" }
 * );
 * ```
 */

import { AgentGuard, ProtectOptions } from "../guard";
import { AgentPolicy } from "../policy";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyFn = (...args: any[]) => any;

export interface LlamaIndexToolOptions {
  /** Override tool name. */
  name?: string;
  /** Fixed cost per call (USD). */
  cost?: number;
  /** Model name for cost tracking. */
  model?: string;
}

/**
 * Wraps LlamaIndex tools and agents with {@link AgentGuard} policy enforcement.
 */
export class LlamaIndexGuard {
  private readonly _guard: AgentGuard;

  constructor(guardOrPolicy: AgentGuard | AgentPolicy) {
    this._guard =
      guardOrPolicy instanceof AgentGuard
        ? guardOrPolicy
        : new AgentGuard({ policy: guardOrPolicy });
  }

  /** Access the underlying {@link AgentGuard}. */
  get guard(): AgentGuard {
    return this._guard;
  }

  /**
   * Wrap a plain function that will be used as a LlamaIndex tool.
   *
   * @param fn   The function to protect.
   * @param opts Optional overrides for name, cost, and model.
   * @returns    Protected function with the same signature.
   */
  tool<T extends AnyFn>(fn: T, opts: LlamaIndexToolOptions = {}): T {
    const options: ProtectOptions = {
      toolName: opts.name ?? fn.name,
      cost: opts.cost,
      model: opts.model,
    };
    return this._guard.protect(fn, options);
  }

  /**
   * Wrap a LlamaIndex tool object (with a `call` method).
   *
   * @param tool  A LlamaIndex BaseTool instance.
   * @param name  Optional override for the tool name.
   * @returns     The same tool with its `call` method patched.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  wrapTool(tool: any, name?: string): any {
    if (tool == null || typeof tool.call !== "function") return tool;

    const toolName: string =
      name ?? tool.metadata?.name ?? tool.constructor?.name ?? "unknown_tool";
    const originalCall = tool.call.bind(tool) as AnyFn;

    tool.call = this._guard.protect(originalCall, { toolName });
    return tool;
  }

  /**
   * Wrap multiple LlamaIndex tool objects.
   *
   * @param tools Array of LlamaIndex BaseTool instances.
   * @returns     The same array with each tool's `call` method patched.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  wrapTools(tools: any[]): any[] {
    return tools.map((t) => this.wrapTool(t));
  }

  /**
   * Wrap a LlamaIndex QueryEngine with protection.
   *
   * @param engine  A LlamaIndex query engine instance.
   * @param name    Tool name for policy matching and audit logs.
   * @returns       The same engine with its `query` method patched.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  wrapQueryEngine(engine: any, name = "query_engine"): any {
    if (engine == null || typeof engine.query !== "function") return engine;

    const originalQuery = engine.query.bind(engine) as AnyFn;
    engine.query = this._guard.protect(originalQuery, { toolName: name });
    return engine;
  }
}

/**
 * Protect a LlamaIndex agent by wrapping all of its tools.
 *
 * @param agent           A LlamaIndex agent instance.
 * @param guardOrPolicy   {@link AgentGuard} or {@link AgentPolicy} to use.
 * @returns               The same agent with its tools wrapped.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function protectLlamaIndexAgent(agent: any, guardOrPolicy: AgentGuard | AgentPolicy): any {
  const llamaGuard = new LlamaIndexGuard(guardOrPolicy);

  if (Array.isArray(agent?._tools)) {
    agent._tools = llamaGuard.wrapTools(agent._tools);
  } else if (Array.isArray(agent?.tools)) {
    agent.tools = llamaGuard.wrapTools(agent.tools);
  }

  return agent;
}

/**
 * Protect a LlamaIndex QueryEngine.
 *
 * @param engine          A LlamaIndex query engine instance.
 * @param guardOrPolicy   {@link AgentGuard} or {@link AgentPolicy} to use.
 * @param name            Tool name for policy matching (default: `"query_engine"`).
 * @returns               The same engine with its `query` method wrapped.
 */
export function protectQueryEngine(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  engine: any,
  guardOrPolicy: AgentGuard | AgentPolicy,
  name = "query_engine",
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): any {
  const llamaGuard = new LlamaIndexGuard(guardOrPolicy);
  return llamaGuard.wrapQueryEngine(engine, name);
}

File: python/agentsentinel/integrations/llamaindex.py
"""LlamaIndex integration for AgentSentinel.

LlamaIndex is a data framework for LLM applications with tools, agents, and RAG.
This integration wraps LlamaIndex tools and query engines with AgentSentinel protection.

Usage::

    from llama_index.core.tools import FunctionTool
    from llama_index.core.agent import ReActAgent
    from agentsentinel import AgentPolicy, AgentGuard
    from agentsentinel.integrations.llamaindex import protect_agent, LlamaIndexGuard

    policy = AgentPolicy(daily_budget=10.0)
    guard = AgentGuard(policy)

    # Option 1: Protect an agent
    agent = ReActAgent.from_tools(tools, llm=llm)
    protected_agent = protect_agent(agent, guard=guard)

    # Option 2: Protect individual tools
    llama_guard = LlamaIndexGuard(guard)

    @llama_guard.tool
    def query_database(query: str) -> str:
        return db.query(query)

    # Option 3: Wrap existing FunctionTool
    tool = FunctionTool.from_defaults(fn=my_func)
    protected_tool = llama_guard.wrap_tool(tool)

.. note::
    This module does **not** import ``llama_index`` at the module level, so
    ``agentsentinel`` has no hard dependency on LlamaIndex.
"""

from __future__ import annotations

import asyncio
import functools
from typing import Any, Callable, List, Optional, TypeVar, Union

from ..guard import AgentGuard
from ..policy import AgentPolicy

F = TypeVar("F", bound=Callable[..., Any])


# ---------------------------------------------------------------------------
# Core guard class
# ---------------------------------------------------------------------------


class LlamaIndexGuard:
    """Wraps LlamaIndex tools and agents with :class:`.AgentGuard` policy enforcement.

    Parameters
    ----------
    guard:
        A pre-configured :class:`.AgentGuard` instance.
    policy:
        :class:`.AgentPolicy` used to create a guard when *guard* is ``None``.
        Ignored when *guard* is provided.

    Example
    -------
    ::

        from agentsentinel import AgentGuard, AgentPolicy
        from agentsentinel.integrations.llamaindex import LlamaIndexGuard

        policy = AgentPolicy(daily_budget=10.0, model_budgets={"gpt-4o": 5.0})
        llama_guard = LlamaIndexGuard(AgentGuard(policy))

        @llama_guard.tool(model="gpt-4o")
        def query_knowledge_base(query: str) -> str:
            return kb.query(query)
    """

    def __init__(
        self,
        guard: Optional[AgentGuard] = None,
        policy: Optional[AgentPolicy] = None,
    ) -> None:
        if guard is not None:
            self._guard = guard
        elif policy is not None:
            self._guard = AgentGuard(policy)
        else:
            self._guard = AgentGuard(AgentPolicy())

    @property
    def guard(self) -> AgentGuard:
        """Access the underlying :class:`.AgentGuard`."""
        return self._guard

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def tool(
        self,
        func: Optional[F] = None,
        *,
        name: Optional[str] = None,
        cost: Optional[float] = None,
        model: Optional[str] = None,
    ) -> Union[F, Callable[[F], F]]:
        """Decorator to protect a function used as a LlamaIndex tool.

        Can be used with or without arguments::

            @llama_guard.tool
            def search_web(query: str) -> str:
                return search(query)

            @llama_guard.tool(model="gpt-4o", cost=0.02)
            def summarize(text: str) -> str:
                return summarize(text)

        Parameters
        ----------
        func:
            The function to protect (when used without parentheses).
        name:
            Override tool name (defaults to function name).
        cost:
            Fixed cost per call (USD).
        model:
            Model name for per-model budget tracking.
        """
        def decorator(f: F) -> F:
            tool_name = name or f.__name__
            return self._guard.protect(f, tool_name=tool_name, cost=cost, model=model)  # type: ignore[return-value]

        if func is not None:
            return decorator(func)
        return decorator

    def wrap_tool(self, tool: Any, name: Optional[str] = None) -> Any:
        """Wrap a LlamaIndex tool (``FunctionTool``, ``QueryEngineTool``, etc.).

        Parameters
        ----------
        tool:
            A LlamaIndex ``BaseTool`` instance.
        name:
            Override tool name (defaults to ``tool.metadata.name``).

        Returns
        -------
        The same tool object with its ``call`` (and ``acall``) patched.
        """
        try:
            from llama_index.core.tools import BaseTool  # type: ignore[import]
        except ImportError:
            return tool

        if not isinstance(tool, BaseTool):
            return tool

        tool_name = name or getattr(getattr(tool, "metadata", None), "name", None) or tool.__class__.__name__
        original_call = tool.call

        @functools.wraps(original_call)
        def protected_call(*args: Any, **kwargs: Any) -> Any:
            protected = self._guard.protect(
                lambda *a, **kw: original_call(*a, **kw),
                tool_name=tool_name,
            )
            return protected(*args, **kwargs)

        tool.call = protected_call  # type: ignore[method-assign]

        # Wrap async variant if present.
        if hasattr(tool, "acall"):
            original_acall = tool.acall

            @functools.wraps(original_acall)
            async def protected_acall(*args: Any, **kwargs: Any) -> Any:
                loop = asyncio.get_running_loop()
                return await loop.run_in_executor(
                    None,
                    lambda: protected_call(*args, **kwargs),
                )

            tool.acall = protected_acall  # type: ignore[method-assign]

        return tool

    def wrap_tools(self, tools: List[Any]) -> List[Any]:
        """Wrap multiple LlamaIndex tools.

        Parameters
        ----------
        tools:
            List of LlamaIndex ``BaseTool`` instances.

        Returns
        -------
        The same list with each tool wrapped.
        """
        return [self.wrap_tool(t) for t in tools]

    def wrap_query_engine(self, engine: Any, name: str = "query_engine") -> Any:
        """Wrap a LlamaIndex ``QueryEngine`` with protection.

        Parameters
        ----------
        engine:
            A LlamaIndex query engine instance.
        name:
            Tool name for policy matching and audit logs.

        Returns
        -------
        The same engine with its ``query`` (and ``aquery``) method patched.
        """
        original_query = engine.query

        @functools.wraps(original_query)
        def protected_query(*args: Any, **kwargs: Any) -> Any:
            protected = self._guard.protect(
                lambda *a, **kw: original_query(*a, **kw),
                tool_name=name,
            )
            return protected(*args, **kwargs)

        engine.query = protected_query  # type: ignore[method-assign]

        if hasattr(engine, "aquery"):
            original_aquery = engine.aquery

            @functools.wraps(original_aquery)
            async def protected_aquery(*args: Any, **kwargs: Any) -> Any:
                loop = asyncio.get_running_loop()
                return await loop.run_in_executor(
                    None,
                    lambda: protected_query(*args, **kwargs),
                )

            engine.aquery = protected_aquery  # type: ignore[method-assign]

        return engine


# ---------------------------------------------------------------------------
# Convenience one-liners
# ---------------------------------------------------------------------------


def protect_agent(
    agent: Any,
    *,
    guard: Optional[AgentGuard] = None,
    policy: Optional[AgentPolicy] = None,
) -> Any:
    """Protect a LlamaIndex agent (``ReActAgent``, ``OpenAIAgent``, etc.).

    Parameters
    ----------
    agent:
        A LlamaIndex agent instance.
    guard:
        Pre-configured :class:`.AgentGuard`.
    policy:
        :class:`.AgentPolicy` used to create a guard when *guard* is ``None``.

    Returns
    -------
    The same agent with its tools wrapped.
    """
    llama_guard = LlamaIndexGuard(guard=guard, policy=policy)

    if hasattr(agent, "_tools"):
        agent._tools = llama_guard.wrap_tools(agent._tools)
    elif hasattr(agent, "tools"):
        agent.tools = llama_guard.wrap_tools(agent.tools)

    return agent


def protect_query_engine(
    engine: Any,
    *,
    guard: Optional[AgentGuard] = None,
    policy: Optional[AgentPolicy] = None,
    name: str = "query_engine",
) -> Any:
    """Protect a LlamaIndex ``QueryEngine``.

    Parameters
    ----------
    engine:
        A LlamaIndex query engine instance.
    guard:
        Pre-configured :class:`.AgentGuard`.
    policy:
        :class:`.AgentPolicy` used to create a guard when *guard* is ``None``.
    name:
        Tool name for policy matching and audit logs.

    Returns
    -------
    The same engine with its ``query`` method wrapped.
    """
    llama_guard = LlamaIndexGuard(guard=guard, policy=policy)
    return llama_guard.wrap_query_engine(engine, name=name)
